staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -132,26 +150,135 @@ func NewCloudNodeController(
+		// put back on the workqueue and attempted again after a back-off
+		// period.
+		defer cnc.workqueue.Done(obj)
+		var key string
+		var ok bool
+		// We expect strings to come off the workqueue. These are of the
+		// form namespace/name. We do this as the delayed nature of the
+		// workqueue meafns the items in the informer cache may actually be
+		// more up to date that when the item was initially put onto the
+		// workqueue.
+		if key, ok = obj.(string); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			cnc.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// Node resource to be synced.
+		if err := cnc.syncHandler(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			cnc.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		cnc.workqueue.Forget(obj)
+		klog.Infof("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler implements the logic of the controller. It compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Node resource with the current status of the resource.
+func (cnc *CloudNodeController) syncHandler(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the Node resource with this namespace/name
+	node, err := cnc.nodesLister.Get(name)
+	if err != nil {
+		// The Node resource may no longer exist, in which case we stop
+		// processing.
+		if apierrors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("Node '%s' in work queue no longer exists", key))
+			return nil
+		}
+
+		return err
+	}
+
+	// Finally, we update the status block of the Node resource to reflect the
+	// current state of the world
+	err = cnc.UpdateNodeStatus(context.TODO())
+	if err != nil {
+		return err
+	}
+
+	cnc.recorder.Event(node, v1.EventTypeNormal, NodeSuccessSynced, NodeMessageResourceSynced)
+	return nil
 }
 
 // UpdateNodeStatus updates the node status, such as node addresses
-func (cnc *CloudNodeController) UpdateNodeStatus(ctx context.Context) {
+func (cnc *CloudNodeController) UpdateNodeStatus(ctx context.Context) error {
 	nodes, err := cnc.kubeClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{ResourceVersion: "0"})
```

Comment:
Is it necessary for us to still call kubeclient api to list nodes?  we can just get all nodes info from the cache(nodelisterner).

Reply:
> Is it necessary for us to still call kubeclient api to list nodes? we can just get all nodes info from the cache(nodelisterner).

Don't think so, we can probably update this to list nodes from the informer cache as well.